tests/bson/bson-enum-002.phpt

```diff
@@ -0,0 +1,104 @@
+--TEST--
+BSON encoding: Enums implementing Unserializable
+--SKIPIF--
+<?php require __DIR__ . "/../utils/basic-skipif.inc"; ?>
+<?php skip_if_php_version('<', '8.1.0'); ?>
+--FILE--
+<?php
+
+require_once __DIR__ . '/../utils/basic.inc';
+
+enum MyEnum implements MongoDB\BSON\Unserializable
+{
+    case foo;
+
+    public function bsonUnserialize(array $data)
+    {
+        /* Enums do not maintain state, so this method serves no practical
+         * purpose other than being required by the interface. Since an
+         * implementation is required, we will log a message to assert that BSON
+         * decoding always invokes this method. */
+        printf("%s called with: %s\n", __METHOD__, json_encode($data));
+    }
+}
+
+enum MyBackedEnum: string implements MongoDB\BSON\Unserializable
+{
+    case foo = 'bar';
+
+    public function bsonUnserialize(array $data)
+    {
+        printf("%s called with: %s\n", __METHOD__, json_encode($data));
+    }
+}
+
+$tests = [
+    /* There is no practical reason to use an enum as a root document, since it
+     * cannot have an "_id" field, but we'll test this anyway since we're only
+     * using BSON functions and not round-tripping data through the server. */
+    [
+        MyEnum::foo,
+        ['root' => MyEnum::class],
+    ],
+    [
+        MyBackedEnum::foo,
+        ['root' => MyBackedEnum::class],
+    ],
+    [
+        ['myEnum' => MyEnum::foo],
+        ['fieldPaths' => ['myEnum' => MyEnum::class]],
+    ],
+    [
+        ['myBackedEnum' => MyBackedEnum::foo],
+        ['fieldPaths' => ['myBackedEnum' => MyBackedEnum::class]],
+    ],
+];
+
+foreach ($tests as $test) {
+    [$document, $typeMap] = $test;
+
+    $bson = fromPHP($document);
+    echo "Test ", toJSON($bson), "\n";
+    hex_dump($bson);
+    var_dump(toPHP($bson, $typeMap));
+    echo "\n";
+}
+
+?>
+===DONE===
```

Comment:
When executed with `TEST_PHP_ARGS=-m`, both `bson-enum-002.phpt` and `bson-enum-003.phpt` fail the following assertion:

```
php: /tmp/build_php-8.1.8.FWj/php-8.1.8/Zend/zend_objects_API.c:195: zend_objects_store_del: Assertion `(!(((zend_uintptr_t)((executor_globals.objects_store).object_buckets[handle])) & (1<<0)))' failed.
Aborted

Termsig=6
```

The corresponding `.mem` file reports many leaks (none include `phongo_bson.c` in the stack trace), but I believe those are just because of the abort failure immediately terminating the process without allowing PHP to clean up internally allocated memory.

The assertion failure corresponds to [this line in LXR](https://heap.space/xref/PHP-8.1/Zend/zend_objects_API.c?r=fdc22744#195).

Reply:
I spent some time investigating this and realized it was caused by using `ZVAL_COPY_VALUE` instead of `ZVAL_COPY` on the resolved enum case. These are discussed in more detail in [Copying Zvals](https://www.phpinternalsbook.com/php7/zvals/memory_management.html#copying-zvals) in the internals book.

 `ZVAL_COPY_VALUE` was previously used on the return value from `bsonSerialize()`, which made sense there since we didn't need to increment the ref count (the value was being moved). But for enums, this left the original enum case in a bad state, which triggered the assertion failure in `zend_objects_store_del()`. `ZVAL_COPY` correctly increments the case's refcount, which allows all tests to pass w/o leaks.